### PR TITLE
Rewrite vload and vstore builtins

### LINKED ISSUE #232: vloadn/vstoren produces incorrect code when addresses aren't aligned to the vector type
The approach taken by `replace{Vload,Vstore}` in the `ReplaceOpenCLBuiltinPass` pass is to cast the pointer to a pointer to the vector type and use a load/store instruction. This interacts badly with `SimplifyPointerBitcastPass::runOnBitcastFromGEP` resulting in the address being rounded down and aligned to the vector size.

### PR DESCRIPTION
**Pull request by alan-baker**
Fixes #232

* Simplify vload and vstore generation by using scalar loads and stores
  * avoid pointer casting, but should still be simple for a driver to
  coalesce if aligned
  * test all supported variants
  * also tested using amber

Note: the tests are auto-generated from c++. If it is desirable, I could add the vec8 and vec16 variants since the pass handles them.

### CODE REVIEW

File: lib/ReplaceOpenCLBuiltinPass.cpp [hunk 1]
```diff
@@ -1902,58 +1902,74 @@ bool ReplaceOpenCLBuiltinPass::replaceMadandMad24andMul24(Module &M) {
...
       SmallVector<Instruction *, 4> ToRemoves;
 
-      // Walk the users of the function.
+      auto fname = F->getName();
+      if (!fname.consume_front("_Z"))
+        continue;
+      size_t name_len;
+      if (fname.consumeInteger(10, name_len))
+        continue;
+      std::string name = fname.take_front(name_len);
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply lib/ReplaceOpenCLBuiltinPass.cpp]: I tried that at first, but It wasn't clear how to best extend it to support parsing pointer mangling. In the end, I don't need to extract the types from the arguments based on the name so I avoided that.

File: lib/ReplaceOpenCLBuiltinPass.cpp [hunk 2]
```diff
@@ -1902,58 +1902,74 @@ bool ReplaceOpenCLBuiltinPass::replaceMadandMad24andMul24(Module &M) {
...
-          // The index argument from vstoren.
-          auto Arg1 = CI->getOperand(1);
+          auto data_type = data->getType();
+          if (!data_type->isVectorTy())
+            continue;
 
-          // The pointer argument from vstoren.
-          auto Arg2 = CI->getOperand(2);
+          auto elems = data_type->getVectorNumElements();
+          if (elems != 2 && elems != 4 && elems != 8 && elems != 16)
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply lib/ReplaceOpenCLBuiltinPass.cpp]: I think that would generate a parser error which is why I didn't check it.

File: lib/ReplaceOpenCLBuiltinPass.cpp [hunk 3]
```diff
@@ -1902,58 +1902,74 @@ bool ReplaceOpenCLBuiltinPass::replaceMadandMad24andMul24(Module &M) {
...
-    auto Name = Op.name;
-    auto N = Op.n;
-    auto TypeFn = Op.get_scalar_type_function;
-    if (auto F = M.getFunction(Name)) {
+  for (auto const &SymVal : M.getValueSymbolTable()) {
+    if (!SymVal.getKey().contains("vstore"))
+      continue;
+    if (SymVal.getKey().contains("vstore_"))
+      continue;
+    if (SymVal.getKey().contains("vstorea"))
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply lib/ReplaceOpenCLBuiltinPass.cpp]: Yes, I believe it's faster than performing all the mangling and switching below.

File: lib/ReplaceOpenCLBuiltinPass.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kpet) [reply lib/ReplaceOpenCLBuiltinPass.cpp]: Yes, you're right, it's useless to check.

**Pull request merged**